File: research/marker-hit-area-investigation/flutter_app/lib/markers/bitmap_generator.dart
import 'dart:typed_data';
import 'dart:ui' as ui;

import 'package:flutter/material.dart';

import 'marker_specs.dart';

/// 形状仕様から PNG バイトを生成する。
/// 透明領域の比率は仕様書 §4.1 と一致するよう描画する。
/// 全形状とも色は固定で red 系（仮説検証に色は無関係）。
class BitmapGenerator {
  /// PNG バイトを返す。
  static Future<Uint8List> generate(MarkerSpec spec) async {
    final w = spec.bitmapPxWidth;
    final h = spec.bitmapPxHeight;

    final recorder = ui.PictureRecorder();
    final canvas = Canvas(recorder);

    // 背景は完全透明（PNG の透明領域として保持）
    final paint = Paint()..style = PaintingStyle.fill;

    switch (spec.shape) {
      case MarkerShape.s0SolidSquare:
        paint.color = const Color(0xFFE53935); // red 600
        canvas.drawRect(Rect.fromLTWH(0, 0, w.toDouble(), h.toDouble()), paint);
        break;

      case MarkerShape.s1Circle:
        paint.color = const Color(0xFFE53935);
        canvas.drawCircle(
          Offset(w / 2, h / 2),
          w / 2,
          paint,
        );
        break;

      case MarkerShape.s2CirclePad20:
        // 周囲 20% パディング → 円の半径は w * 0.5 * 0.6 = w*0.3
        paint.color = const Color(0xFFE53935);
        canvas.drawCircle(
          Offset(w / 2, h / 2),
          w * 0.3,
          paint,
        );
        break;

      case MarkerShape.s3CirclePad40:
        // 周囲 40% パディング → 円の半径は w * 0.5 * 0.2 = w*0.1
        paint.color = const Color(0xFFE53935);
        canvas.drawCircle(
          Offset(w / 2, h / 2),
          w * 0.1,
          paint,
        );
        break;

      case MarkerShape.s4PinTight:
        // 円(上半分) + 下向きポインタ。bbox はちょうどフィット
        // bitmapPxHeight は w * 1.3 になっている前提
        _drawPin(canvas, w.toDouble(), h.toDouble(), padding: 0.0);
        break;

      case MarkerShape.s5PinPad10:
        // 正方形 bbox（ここでは w × w）+ 周囲 10% パディング
        // 元のピン形状は w*0.8 × w*0.8*1.3 で中央に配置
        _drawPin(canvas, w.toDouble(), h.toDouble(), padding: 0.1);
        break;
    }

    final picture = recorder.endRecording();
    final image = await picture.toImage(w, h);
    final byteData = await image.toByteData(format: ui.ImageByteFormat.png);
    return byteData!.buffer.asUint8List();
  }

  static void _drawPin(Canvas canvas, double w, double h, {required double padding}) {
    final paint = Paint()
      ..color = const Color(0xFFE53935)
      ..style = PaintingStyle.fill;

    // padding は片側比率。実領域は (w - 2*pad*w) × (h - 2*pad*h)
    final padX = w * padding;
    final padY = h * padding;
    final innerW = w - 2 * padX;
    final innerH = h - 2 * padY;

    // 円の半径 = innerW / 2、中心 = (innerW/2, innerW/2)（円は上に配置）
    final circleRadius = innerW / 2;
    final cx = padX + innerW / 2;
    final cy = padY + circleRadius;

    canvas.drawCircle(Offset(cx, cy), circleRadius, paint);

    // ポインタは下向き三角形：円の底（cy + r）から bottom (padY + innerH) まで
    final path = Path()
      ..moveTo(cx - circleRadius * 0.5, cy + circleRadius * 0.6)
      ..lineTo(cx + circleRadius * 0.5, cy + circleRadius * 0.6)
      ..lineTo(cx, padY + innerH)
      ..close();
    canvas.drawPath(path, paint);
  }
}

File: research/marker-hit-area-investigation/android_native_app/app/src/main/kotlin/com/example/markerhit/nativeapp/markers/BitmapGenerator.kt
package com.example.markerhit.nativeapp.markers

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Path
import android.util.DisplayMetrics

/**
 * 形状仕様から Bitmap を生成。Flutter 側 `bitmap_generator.dart` と同じ幾何。
 * density による論理サイズ調整は呼び出し側で `Bitmap.setDensity` する。
 */
object BitmapGenerator {

    /** ヒット時のハイライト用に色だけ差し替えた版を生成 */
    fun generateHighlighted(spec: MarkerSpec): Bitmap =
        generate(spec, highlightColor = "#1B5E20") // dark green

    fun generate(spec: MarkerSpec, highlightColor: String? = null): Bitmap {
        val w = spec.bitmapPxWidth
        val h = spec.bitmapPxHeight
        val bmp = Bitmap.createBitmap(w, h, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bmp)
        // 背景は透明
        canvas.drawColor(Color.TRANSPARENT)

        val paint = Paint().apply {
            color = Color.parseColor(highlightColor ?: "#E53935")
            style = Paint.Style.FILL
            isAntiAlias = true
        }

        when (spec.shape) {
            MarkerShape.S0_SOLID_SQUARE -> {
                canvas.drawRect(0f, 0f, w.toFloat(), h.toFloat(), paint)
            }
            MarkerShape.S1_CIRCLE -> {
                canvas.drawCircle(w / 2f, h / 2f, w / 2f, paint)
            }
            MarkerShape.S2_CIRCLE_PAD20 -> {
                canvas.drawCircle(w / 2f, h / 2f, w * 0.3f, paint)
            }
            MarkerShape.S3_CIRCLE_PAD40 -> {
                canvas.drawCircle(w / 2f, h / 2f, w * 0.1f, paint)
            }
            MarkerShape.S4_PIN_TIGHT -> {
                drawPin(canvas, w.toFloat(), h.toFloat(), padding = 0f, paint = paint)
            }
            MarkerShape.S5_PIN_PAD10 -> {
                drawPin(canvas, w.toFloat(), h.toFloat(), padding = 0.1f, paint = paint)
            }
        }

        // density 設定で論理サイズを制御。
        // R1: DENSITY_DEFAULT (160 = 1x) → 32px bitmap が 32 dp で表示
        // R3: 480 (3x) → 96px bitmap が 32 dp で表示
        // RN: DENSITY_NONE → Maps SDK は scale せず raw px を使う想定
        when (spec.ratio) {
            MarkerRatio.R1 -> bmp.density = DisplayMetrics.DENSITY_DEFAULT
            MarkerRatio.R3 -> bmp.density = 480
            MarkerRatio.RN -> bmp.density = Bitmap.DENSITY_NONE
        }
        return bmp
    }

    private fun drawPin(canvas: Canvas, w: Float, h: Float, padding: Float, paint: Paint) {
        val padX = w * padding
        val padY = h * padding
        val innerW = w - 2 * padX
        val innerH = h - 2 * padY
        val circleRadius = innerW / 2f
        val cx = padX + innerW / 2f
        val cy = padY + circleRadius
        canvas.drawCircle(cx, cy, circleRadius, paint)
        val path = Path().apply {
            moveTo(cx - circleRadius * 0.5f, cy + circleRadius * 0.6f)
            lineTo(cx + circleRadius * 0.5f, cy + circleRadius * 0.6f)
            lineTo(cx, padY + innerH)
            close()
        }
        canvas.drawPath(path, paint)
    }
}
